lightning/src/events/mod.rs

```diff
@@ -67,6 +68,8 @@ pub enum PaymentPurpose {
 		/// [`ChannelManager::create_inbound_payment`]: crate::ln::channelmanager::ChannelManager::create_inbound_payment
 		/// [`ChannelManager::create_inbound_payment_for_hash`]: crate::ln::channelmanager::ChannelManager::create_inbound_payment_for_hash
 		payment_secret: PaymentSecret,
+		///
+		payment_context: Option<PaymentContext>,
```

Comment:
Initially, my intention was to have a top-level field in `PaymentClaimable`. But while piping that data through I noticed that the approach here seemed to make sense internally at least. We have `ClaimingPayment` and `ClaimablePayment`, both of which have a required `PaymentPurpose` field. So while I could make another variant to `OnionPayload` for `PaymentContext`, we'd still need a `PaymentPurpose` to create those structs (`ClaimingPayment` and `ClaimablePayment`). So this seemed like the least disruptive approach, but definitely looking for a Concept ACK on this or an alternative approach.

Reply:
Ah, let me give that a try.

Reply:
Oh, `PaymentPurpose` doesn't use `impl_writeable_tlv_based_enum_upgradable`. Would that be a problem?

Reply:
> Ugh, kinda (of course even if it did we'd still be bitten by #2969)...I mean we should try to get the API right, I think, though, even if that means we upgrade the read-side logic in this release and then say you can't downgrade past the current release if you use BOLT12 in the next.

I can add a `Bolt12InvoicePayment` variant though it's gonna have the same fields as `InvoicePayment` plus an optional (still) `PaymentContext` because we want to support downgrades there. I'm fine with that, but maybe it makes more sense to rename `InvoicePayment` to `ExpectedPayment` (opposite to the `SpontaneousPayment` variant) and add an `Option<PaymentContext>` as is currently done here?

That way we don't have two levels of payment types (i.e. `PaymentPurpose` and `PaymentContext`). And we don't break downgrades that way. Thoughts?

> That said, `PaymentClaimable` already writes the individual parts in TLVs so we could still figure out how to handle downgrade there. `PaymentClaimed` is a bit more awkward, but we could serialize a compat shim (somewhat funnily, it already reads via an `upgradable_required` as if it were upgradable, but its not...).

Hmm... we do serialize `PaymentPurpose` in `ChannelManager`, though, as a whole.

Reply:
> Not sure I understand this, can't we have it be required and read an `upgradable_required` and then support downgrades by discarding the event?

Yeah, I think you're right.

> 
> > I'm fine with that, but maybe it makes more sense to rename InvoicePayment to ExpectedPayment (opposite to the SpontaneousPayment variant) and add an Option as is currently done here?
> 
> Seems fine with me.

Ok, I'll try this out. 

> I wonder if we shouldn't make `PaymentContext` non-Option and have an empty BOLT11 variant, then?

Same here.

Reply:
FYI, still need to finish this.

Reply:
The latest push makes `PaymentContext` non-`Option` in all structs, which means downgrades will fail for unknown contexts. It also uses a `Bolt11Invoice` context for any existing serializations without one. @TheBlueMatt let me know if this is what you had in mind.